### PR TITLE
`sparse_matrix` support for `Sum` and `Prod`

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
We would like all arithmetic operators to support the `sparse_matrix()` method in order to improve overall performance. 

**Description of the Change:**
Use the `sparse_expand_matrix` method (introduced [here](https://github.com/PennyLaneAI/pennylane/pull/2998)) in to expand and combine factors / summands which act on different wires.

We now support: 

```
>>> XZ = qml.prod(qml.PauliX(0), qml.PauliZ(1))
>>> my_op = qml.op_sum(XZ, qml.Hadamard(1))
>>>
>>> print(my_op.sparse_matrix(wire_order=[0,1]))
  (0, 0)	0.7071067811865475
  (0, 1)	0.7071067811865475
  (0, 2)	1.0
  (1, 0)	0.7071067811865475
  (1, 1)	-0.7071067811865475
  (1, 3)	-1.0
  (2, 0)	1.0
  (2, 2)	0.7071067811865475
  (2, 3)	0.7071067811865475
  (3, 1)	-1.0
  (3, 2)	0.7071067811865475
  (3, 3)	-0.7071067811865475
```

### CODE REVIEW

File: pennylane/operation.py [hunk 1]
```diff
@@ -232,6 +232,95 @@ def expand_matrix(base_matrix, wires, wire_order=None):
...
+
+                temp = expanded_wires[i]
+                expanded_wires[i] = expanded_wires[j]
+                expanded_wires[j] = temp
+                break
+
+    expanded_matrix = csr_matrix(U.T) @ expanded_matrix @ U
+    return expanded_matrix
+
+
```
Comment (Jaybsoni) [inline pennylane/operation.py:323]: Same changes as in the PR [here](https://github.com/PennyLaneAI/pennylane/pull/2998)

Comment (Jaybsoni) [reply pennylane/operation.py:323]: We should wait till that one gets approved before merging this one !

**Comment by Jaybsoni** [comment]
[sc-22458]

File: tests/test_operation.py [hunk 2]
```diff
@@ -1796,8 +1796,6 @@ def test_terms_undefined(self):
 
     def test_sparse_matrix_undefined(self):
         """Tests that custom error is raised in the default sparse matrix representation."""
-        with pytest.raises(NotImplementedError):
-            MyOp(wires="a").sparse_matrix(wire_order=["a", "b"])
```
Comment (Jaybsoni) [inline tests/test_operation.py:1800]: We now support `wire_order` in `sparse_matrix()`!

File: pennylane/operation.py (hunk 1 above)
Comment (Jaybsoni) [reply pennylane/operation.py:323]: Code cov expected to fail because of this

File: pennylane/operation.py [hunk 3]
```diff
@@ -521,9 +521,6 @@ def sparse_matrix(self, wire_order=None):
         operator's wires in the global wire order. Otherwise, the wire order defaults to the
         operator's wires.
 
-        .. note::
```
Comment (albi3ro) [inline pennylane/operation.py:524]: Always wonderful to remove things like this!

File: pennylane/operation.py [hunk 4]
```diff
@@ -535,12 +532,14 @@ def sparse_matrix(self, wire_order=None):
...
         """
-        if wire_order is not None:
-            raise NotImplementedError("The wire_order argument is not yet implemented")
         canonical_sparse_matrix = self.compute_sparse_matrix(
             *self.parameters, **self.hyperparameters
         )
-        return canonical_sparse_matrix
+
+        if wire_order is None or self.wires == Wires(wire_order):
+            return canonical_sparse_matrix
```
Comment (albi3ro) [inline pennylane/operation.py:540]: `expand_matrix` now performs this check itself, so we can remove these lines.

```suggestion
```

File: pennylane/ops/op_math/prod.py [hunk 5]
```diff
@@ -300,6 +300,14 @@ def matrix(self, wire_order=None):
         )
         return reduce(math.dot, mats)
 
+    def sparse_matrix(self, wire_order=None):
+        """Compute the sparse matrix representation of the Prod op in csr representation."""
+        if wire_order is None:
+            wire_order = self.wires
```
Comment (albi3ro) [inline pennylane/ops/op_math/prod.py:306]:
```suggestion
        wire_order = wire_order or self.wires
```

File: pennylane/ops/op_math/prod.py [hunk 6]
```diff
@@ -300,6 +300,14 @@ def matrix(self, wire_order=None):
...
+
```
Comment (Jaybsoni) [inline pennylane/ops/op_math/prod.py:307]:
```suggestion
```

File: pennylane/ops/op_math/sum.py [hunk 7]
```diff
@@ -305,6 +305,14 @@ def matrix_gen(summands, wire_order=None):
 
         return _sum(matrix_gen(self.summands, wire_order))
 
+    def sparse_matrix(self, wire_order=None):
+        """Compute the sparse matrix representation of the Sum op in csr representation."""
+        if wire_order is None:
+            wire_order = self.wires
+
+        mats_gen = (op.sparse_matrix(wire_order=wire_order) for op in self.summands)
```
Comment (Jaybsoni) [inline pennylane/ops/op_math/sum.py:313]:
```suggestion
        wire_order = wire_order or self.wires
        mats_gen = (op.sparse_matrix(wire_order=wire_order) for op in self.summands)
```

**Review by AlbertMitjans — APPROVED** [review]
Amazing!! 💯 🚀 

Would be nice to add some tests with sums and prods of `Adjoint` and `Pow` operators.

File: tests/math/test_matrix_manipulation.py [hunk 8]
```diff
@@ -530,6 +530,30 @@ class TestExpandMatrixSparse:
...
+        res.sort_indices()
+        expected = csr_matrix(np.array([[0, 0, 1, 0], [0, 0, 0, 1], [1, 0, 0, 0], [0, 1, 0, 0]]))
+        expected.sort_indices()
+
+        assert type(res) == type(expected)
+        assert all(res.data == expected.data)
+        assert all(res.indices == expected.indices)
+
+    def test_wires_tuple(self):
+        """Tests the case wires is wires.Wires object"""
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply tests/math/test_matrix_manipulation.py:546]: Nice catch!

**Review by albi3ro — APPROVED** [review]
Correct answers give us a good starting point for further performance optimizations. Nice to have the feature coverage 👍

**Pull request merged**